staging/src/k8s.io/code-generator/cmd/api-linter/linters/api-linter.go

```diff
@@ -0,0 +1,258 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package linters
+
+import (
+	"bufio"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+
+	"k8s.io/gengo/args"
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/namer"
+	"k8s.io/gengo/types"
+	"k8s.io/kube-openapi/pkg/util"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// This is the comment tag that carries parameters for OpenAPI generation
+	tagName     = "k8s:openapi-gen"
+	tagOptional = "optional"
+
+	// Known values for the tag
+	tagValueTrue  = "true"
+	tagValueFalse = "false"
+)
+
+// CustomArgs is used by the gengo framework to pass args specific to this linter.
+type CustomArgs struct {
+	// WhitelistFilename is the whitelist csv filename.
+	WhitelistFilename string
+}
+
+func getAPITagValue(comments []string) []string {
+	return types.ExtractCommentTags("+", comments)[tagName]
+}
+
+func hasAPITagValue(comments []string, value string) bool {
+	tagValues := getAPITagValue(comments)
+	for _, val := range tagValues {
+		if val == value {
+			return true
+		}
+	}
+	return false
+}
+
+// hasOptionalTag returns true if the member has +optional in its comments or
+// omitempty in its json tags.
+func hasOptionalTag(m *types.Member) bool {
+	hasOptionalCommentTag := types.ExtractCommentTags(
+		"+", m.CommentLines)[tagOptional] != nil
+	hasOptionalJSONTag := strings.Contains(
+		reflect.StructTag(m.Tags).Get("json"), "omitempty")
+	return hasOptionalCommentTag || hasOptionalJSONTag
+}
+
+// NameSystems returns the name system used by the generators in this package.
+func NameSystems() namer.NameSystems {
+	return namer.NameSystems{
+		"raw": namer.NewRawNamer("", nil),
+	}
+}
+
+// DefaultNameSystem returns the default name system for ordering the types to be
+// processed by the generators in this package.
+func DefaultNameSystem() string {
+	return "raw"
+}
+
+// Packages returns a API linter that validates API conventions and generates a dummy package.
```

Comment:
Hopefully you can find a way of accomplishing this that doesn't actually involve making a new package.

Reply:
Currently `gengo` supports `assemble` and `verify` mode. I think to avoid generate new file/dir we need a third "dryrun" mode. I will open a PR for that